src/dmd/target.d

```diff
@@ -741,6 +741,64 @@ extern (C++) struct Target
         return global.params.is64bit ? (sz + 7) & ~7 : (sz + 3) & ~3;
     }
 
+    /**
+     * Determine which `in` parameters needs to be passed by `ref`
+     *
+     * Called from `TypeFunction` semantic with the full function type.
+     * This routine must iterate over parameters, and may set `STC.ref_`
+     * for any parameter which already have `STC.in_`.
+     * This hook must never set `STC.ref_` if the parameter is not `STC.in_`,
+     * nor should it ever change anything else.
+     *
+     * This hook will not be called when `-preview=in` wasn't passed to the
+     * frontend, hence it needs not care about `global.params.previewIn`.
+     *
+     * Params:
+     *   tf    = Type of the function to inspect. The type will have its
+     *           parameter types semantically resolved, however other attributes
+     *           (return type, `@safe` / `nothrow`, etc...) must not be used.
+     */
+    extern(C++) void applyInRefParams (TypeFunction tf)
+    {
+        if (!global.params.previewIn)
+            return;
+
+        foreach (_idx, p; tf.parameterList)
+        {
+            // Ignore non-`in` or already-`ref` parameters
+            if ((p.storageClass & (STC.in_ | STC.ref_)) != STC.in_)
+                continue;
+
+            // If it has a copy constructor / destructor / postblit,
+            // it is always by ref
+            if (p.type.needsDestruction() || p.type.needsCopyOrPostblit())
+                p.storageClass |= STC.ref_;
+            // If the type can't be copied, always by `ref`
+            else if (!p.type.isCopyable())
+                p.storageClass |= STC.ref_;
+            // The Win64 ABI requires x87 real to be passed by ref
+            else if (p.type.ty == Tfloat80)
+                p.storageClass |= STC.ref_;
+
+            // If it's a dynamic array, use the value type as it
+            // allows covariance between `in char[]` and `scope const(char)[]`
+            // The same reasoning applies to pointers and classes,
+            // but that is handled by the `(sz > 8)` below.
+            else if (p.type.ty == Tarray)
+                continue;
+            // Pass delegates by value to allow covariance
+            // Function pointers are a single pointers and handled below.
+            else if (p.type.ty == Tdelegate)
+                continue;
+            else
+            {
+                const sz = p.type.size();
+                if (global.params.is64bit ? (sz > 16) : (sz > 8))
```

Comment:
What is this really for?  Passing vectors by reference? Passing large structs by reference?  Why not test whether the type should be returned in memory instead?

Reply:
Large structs and static arrays

Reply:
And is passing `in __vector` by reference intentional?

Reply:
https://github.com/dlang/dmd/pull/11000#issuecomment-674508079

> TBH, regarding the best ABI, I'd appreciate if you (or people more familiar with calling conventions) could follow up on this.

Reply:
For dmd, `target.isReturnOnStack` or `target.toArgTypes` would probably be the best gauge for deciding whether to pass by ref or not.  There's no point passing by ref if its going in registers.

Reply:
That's clear and has been postponed to a follow-up; this PR isn't of much use until then.

Reply:
For DMD, it should be ok to do:
```
else
{
    auto tup = target.toArgTypes(p.type);
    if (!tup.arguments)
        p.storageClass |= STC.ref_;
}
```
Size alone is a bit too arbitrary a gauge.

Reply:
@ibuclaw : Thanks, done

Reply:
As shown by CI, this doesn't work for Win64 at all (returns null), *and* is costly for structs, where the argTypes are cached as mentioned earlier.